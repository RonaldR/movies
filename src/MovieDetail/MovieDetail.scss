.movie-detail {
	width: 100%;
	max-width: 640px;
	margin: 0 auto;
	background: #fafafa;
	padding: 18px;
}

.movie-detail__details {
	display: flex;
	flex-wrap: wrap;

	.poster {
		width: 100%;
		max-width: 300px;
		margin-right: 12px;

		img {
			width: 100%;
		}
	}
}

.movie-detail__list {
	margin: 0;
	padding: 0;
	list-style: none;
	flex: 1;
}

.message,
.back-link {
	font-size: 24px;
	color: white;
	text-align: center;
	display: block;
	margin: 16px;
	text-decoration: none;
}

.back-link {
	.back-link__icon {
		transition: 0.2s;
		display: inline-block;
	}

	&:hover .back-link__icon {
		transform: translateX(-5px);
	}
}
