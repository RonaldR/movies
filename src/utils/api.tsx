const apiKey = '26fc32c3'; // should be in .env

export const searchMovie = async (search: string) => {
    const response = await fetch(`https://www.omdbapi.com/?s=${search}&apikey=${apiKey}`);
    return await response.json();
}

export const getMovie = async (movieId: string) => {
    const response = await fetch(`https://www.omdbapi.com/?i=${movieId}&apikey=${apiKey}`);
    return await response.json();
}