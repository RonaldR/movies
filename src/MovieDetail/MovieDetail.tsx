import React, { useState, useEffect } from 'react';
import { useParams, Link } from 'react-router-dom';
import MovieDetailListItem from '../MovieDetailListItem/MovieDetailListItem';
import { getMovie } from '../utils/api';

import { Movie } from '../types/types';

import './MovieDetail.scss';

const MovieDetail = () => {
    const { movieId } = useParams();
    const [movie, setMovie] = useState<Movie | undefined>()
    const [message, setMessage] = useState('‚è≥');

    useEffect(() => {
        (async () => {
            try {
                const json = await getMovie(movieId);
                if (json) {
                    if (json.Response === 'True') {
                        setMovie(json);
                    } else {
                        setMessage('Movie not found üò¢');
                    }
                } else {
                    setMessage('Error while getting your movie üò¢');
                }
            } catch (error) {
                setMessage('Error while getting your movie üò¢');
            }
        })();

    }, [movieId]);

    return <>
        <article className="movie-detail">
            {movie ? (
                <>
                    <h1>{movie.Title} ({movie.Year})</h1>
                    <p>{movie.Plot}</p>

                    <div className="movie-detail__details">
                        <figure className="poster">
                            <img src={movie.Poster} alt="" />
                        </figure>

                        <ul className="movie-detail__list">
                            <MovieDetailListItem label="Rating" item={`${movie.imdbRating} (IMDB)`} link={`https://www.imdb.com/title/${movie.imdbID}/`} />
                            <MovieDetailListItem label="Metascore" item={movie.Metascore} />
                            <MovieDetailListItem label="Director(s)" item={movie.Director} />
                            <MovieDetailListItem label="Writer(s)" item={movie.Writer} />
                            <MovieDetailListItem label="Actors" item={movie.Actors} />
                            <MovieDetailListItem label="Genre" item={movie.Genre} />
                            <MovieDetailListItem label="Rated" item={movie.Rated} />
                            <MovieDetailListItem label="Runtime" item={movie.Runtime} />
                            {movie.BoxOffice !== 'N/A' && <MovieDetailListItem label="Money" item={movie.BoxOffice} />}
                            <MovieDetailListItem label="Awards" item={movie.Awards} />
                        </ul>
                    </div>
                </>)
                : <p className="message">{message}</p>}
        </article>

        <Link to="/" className="back-link">
            <span className="back-link__icon" role="img" aria-label="arrow-back">‚¨ÖÔ∏è</span> Back to searching for more
        </Link>
    </>;
}


export default MovieDetail;
