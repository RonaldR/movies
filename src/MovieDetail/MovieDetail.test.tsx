import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import MovieDetail from './MovieDetail';

import { getMovie } from '../utils/api';

jest.mock('../utils/api');
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
    useParams: () => ({
        movieId: 'tt0133093',
    }),
}));

beforeEach(() => {
    jest.resetAllMocks();
});

test('movie detail success', async () => {
    // Matrix
    getMovie.mockResolvedValueOnce({ "Title": "The Matrix", "Year": "1999", "Rated": "R", "Released": "31 Mar 1999", "Runtime": "136 min", "Genre": "Action, Sci-Fi", "Director": "Lana Wachowski, Lilly Wachowski", "Writer": "Lilly Wachowski, Lana Wachowski", "Actors": "Keanu Reeves, Laurence Fishburne, Carrie-Anne Moss, Hugo Weaving", "Plot": "A computer hacker learns from mysterious rebels about the true nature of his reality and his role in the war against its controllers.", "Language": "English", "Country": "USA", "Awards": "Won 4 Oscars. Another 37 wins & 50 nominations.", "Poster": "https://m.media-amazon.com/images/M/MV5BNzQzOTk3OTAtNDQ0Zi00ZTVkLWI0MTEtMDllZjNkYzNjNTc4L2ltYWdlXkEyXkFqcGdeQXVyNjU0OTQ0OTY@._V1_SX300.jpg", "Ratings": [{ "Source": "Internet Movie Database", "Value": "8.7/10" }, { "Source": "Rotten Tomatoes", "Value": "87%" }, { "Source": "Metacritic", "Value": "73/100" }], "Metascore": "73", "imdbRating": "8.7", "imdbVotes": "1,597,196", "imdbID": "tt0133093", "Type": "movie", "DVD": "21 Sep 1999", "BoxOffice": "N/A", "Production": "Warner Bros. Pictures", "Website": "N/A", "Response": "True" });

    const { asFragment, getByText } = render(<MemoryRouter initialEntries={["/detail/tt0133093"]}><MovieDetail /></MemoryRouter>);
    // empty state
    expect(asFragment()).toMatchSnapshot();

    await waitFor(() => {
        expect(getByText(/the matrix/i)).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
    expect(getMovie).toHaveBeenCalledTimes(1);
    expect(getMovie).toHaveBeenCalledWith('tt0133093');
});

test('movie detail fetch response false', async () => {
    getMovie.mockResolvedValueOnce({ "Response": "False" });

    const { asFragment, getByText } = render(<MemoryRouter initialEntries={["/detail/tt0133093"]}><MovieDetail /></MemoryRouter>);

    await waitFor(() => {
        expect(getByText(/Movie not found/i)).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
    expect(getMovie).toHaveBeenCalledTimes(1);
});

test('movie detail fetch empty', async () => {
    getMovie.mockResolvedValueOnce(null);

    const { asFragment, getByText } = render(<MemoryRouter initialEntries={["/detail/tt0133093"]}><MovieDetail /></MemoryRouter>);

    await waitFor(() => {
        expect(getByText(/Error while getting your movie/i)).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
    expect(getMovie).toHaveBeenCalledTimes(1);
});