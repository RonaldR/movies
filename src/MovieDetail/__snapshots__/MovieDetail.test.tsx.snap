// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`movie detail fetch empty 1`] = `
<DocumentFragment>
  <article
    class="movie-detail"
  >
    <p
      class="message"
    >
      Error while getting your movie üò¢
    </p>
  </article>
  <a
    class="back-link"
    href="/"
  >
    <span
      aria-label="arrow-back"
      class="back-link__icon"
      role="img"
    >
      ‚¨ÖÔ∏è
    </span>
     Back to searching for more
  </a>
</DocumentFragment>
`;

exports[`movie detail fetch response false 1`] = `
<DocumentFragment>
  <article
    class="movie-detail"
  >
    <p
      class="message"
    >
      Movie not found üò¢
    </p>
  </article>
  <a
    class="back-link"
    href="/"
  >
    <span
      aria-label="arrow-back"
      class="back-link__icon"
      role="img"
    >
      ‚¨ÖÔ∏è
    </span>
     Back to searching for more
  </a>
</DocumentFragment>
`;

exports[`movie detail success 1`] = `
<DocumentFragment>
  <article
    class="movie-detail"
  >
    <p
      class="message"
    >
      ‚è≥
    </p>
  </article>
  <a
    class="back-link"
    href="/"
  >
    <span
      aria-label="arrow-back"
      class="back-link__icon"
      role="img"
    >
      ‚¨ÖÔ∏è
    </span>
     Back to searching for more
  </a>
</DocumentFragment>
`;

exports[`movie detail success 2`] = `
<DocumentFragment>
  <article
    class="movie-detail"
  >
    <h1>
      The Matrix (1999)
    </h1>
    <p>
      A computer hacker learns from mysterious rebels about the true nature of his reality and his role in the war against its controllers.
    </p>
    <div
      class="movie-detail__details"
    >
      <figure
        class="poster"
      >
        <img
          alt=""
          src="https://m.media-amazon.com/images/M/MV5BNzQzOTk3OTAtNDQ0Zi00ZTVkLWI0MTEtMDllZjNkYzNjNTc4L2ltYWdlXkEyXkFqcGdeQXVyNjU0OTQ0OTY@._V1_SX300.jpg"
        />
      </figure>
      <ul
        class="movie-detail__list"
      >
        <li
          class="list-item"
        >
          <span
            class="list-item__label"
          >
            Rating
          </span>
          <span
            class="list-item__item"
          >
            <a
              href="https://www.imdb.com/title/tt0133093/"
              rel="noopener noreferrer"
              target="_blank"
            >
              8.7 (IMDB)
            </a>
          </span>
        </li>
        <li
          class="list-item"
        >
          <span
            class="list-item__label"
          >
            Metascore
          </span>
          <span
            class="list-item__item"
          >
            73
          </span>
        </li>
        <li
          class="list-item"
        >
          <span
            class="list-item__label"
          >
            Director(s)
          </span>
          <span
            class="list-item__item"
          >
            Lana Wachowski, Lilly Wachowski
          </span>
        </li>
        <li
          class="list-item"
        >
          <span
            class="list-item__label"
          >
            Writer(s)
          </span>
          <span
            class="list-item__item"
          >
            Lilly Wachowski, Lana Wachowski
          </span>
        </li>
        <li
          class="list-item"
        >
          <span
            class="list-item__label"
          >
            Actors
          </span>
          <span
            class="list-item__item"
          >
            Keanu Reeves, Laurence Fishburne, Carrie-Anne Moss, Hugo Weaving
          </span>
        </li>
        <li
          class="list-item"
        >
          <span
            class="list-item__label"
          >
            Genre
          </span>
          <span
            class="list-item__item"
          >
            Action, Sci-Fi
          </span>
        </li>
        <li
          class="list-item"
        >
          <span
            class="list-item__label"
          >
            Rated
          </span>
          <span
            class="list-item__item"
          >
            R
          </span>
        </li>
        <li
          class="list-item"
        >
          <span
            class="list-item__label"
          >
            Runtime
          </span>
          <span
            class="list-item__item"
          >
            136 min
          </span>
        </li>
        <li
          class="list-item"
        >
          <span
            class="list-item__label"
          >
            Awards
          </span>
          <span
            class="list-item__item"
          >
            Won 4 Oscars. Another 37 wins & 50 nominations.
          </span>
        </li>
      </ul>
    </div>
  </article>
  <a
    class="back-link"
    href="/"
  >
    <span
      aria-label="arrow-back"
      class="back-link__icon"
      role="img"
    >
      ‚¨ÖÔ∏è
    </span>
     Back to searching for more
  </a>
</DocumentFragment>
`;
